File: 01-Languages/Go/10-File-Handling/main.go
// main.go - os text file I/O, encoding/json (built-in), missing-file handling.
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

type Config struct {
	Theme    string `json:"theme"`
	FontSize int    `json:"fontSize"`
}

func main() {
	path := filepath.Join(os.TempDir(), "example-notes.txt")

	fmt.Println("--- text file round-trip ---")
	err := os.WriteFile(path, []byte("Hello, file system!\n"), 0644)
	if err != nil {
		panic(err)
	}
	contents, err := os.ReadFile(path)
	if err != nil {
		panic(err)
	}
	fmt.Print("Read back: ", string(contents))

	fmt.Println("\n--- encoding/json (built into the standard library) ---")
	config := Config{Theme: "dark", FontSize: 14}
	data, err := json.Marshal(config)
	if err != nil {
		panic(err)
	}
	fmt.Println("Marshaled:", string(data))

	var loaded Config
	err = json.Unmarshal(data, &loaded)
	if err != nil {
		panic(err)
	}
	fmt.Printf("Unmarshaled: Theme=%s, FontSize=%d\n", loaded.Theme, loaded.FontSize)

	fmt.Println("\n--- missing file handled via os.IsNotExist ---")
	missingPath := filepath.Join(os.TempDir(), "does-not-exist-example.txt")
	_, err = os.ReadFile(missingPath)
	if os.IsNotExist(err) {
		fmt.Println("File doesn't exist -- using defaults, handled gracefully")
	}

	os.Remove(path)
	fmt.Println("\nCleaned up temporary file.")
}

File: 01-Languages/Rust/10-File-Handling/main.rs
// main.rs - std::fs text file I/O with Result, and the ErrorKind::NotFound pattern.

use std::env;
use std::fs;
use std::io::ErrorKind;
use std::path::PathBuf;

fn main() {
    let mut path = env::temp_dir();
    path.push("example-notes.txt");

    println!("--- text file round-trip ---");
    fs::write(&path, "Hello, file system!\n").expect("failed to write file");
    let contents = fs::read_to_string(&path).expect("failed to read file");
    print!("Read back: {}", contents);

    println!("\n--- missing file handled via ErrorKind::NotFound ---");
    let mut missing_path = PathBuf::from(env::temp_dir());
    missing_path.push("does-not-exist-example.txt");
    match fs::read_to_string(&missing_path) {
        Ok(c) => println!("{}", c),
        Err(e) if e.kind() == ErrorKind::NotFound => {
            println!("File doesn't exist -- using defaults, handled gracefully");
        }
        Err(e) => println!("Unexpected error: {}", e),
    }

    fs::remove_file(&path).expect("failed to clean up");
    println!("\nCleaned up temporary file.");
}
